src/main/java/org/elasticsearch/index/translog/fs/FsChannelImmutableReader.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.translog.fs;
+
+import org.elasticsearch.ElasticsearchIllegalStateException;
+import org.elasticsearch.common.io.Channels;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+/**
+ * a channel reader which is fixed in length
+ */
+public class FsChannelImmutableReader extends FsChannelReader {
+
+    private final int totalOperations;
+    private final long length;
+
+    /**
+     * Create a snapshot of translog file channel. The length parameter should be consistent with totalOperations and point
+     * at the end of the last operation in this snapshot.
+     */
+    public FsChannelImmutableReader(long id, ChannelReference channelReference, long length, int totalOperations) {
+        super(id, channelReference);
+        this.length = length;
+        this.totalOperations = totalOperations;
+    }
+
+
+    public FsChannelImmutableReader immutableReader() {
```

Comment:
I'm unclear here why you would call `immutableReader` if you already had a copy of the `FsChannelImmutableReader`? It seems like it is cloning the current reader and incrementing the ref on the channel reference? If that's the case, maybe `.clone()` would be a better name for it?